hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java

```diff
@@ -371,27 +350,50 @@ public void deleteByControllerID(final String controllerID) {
 
-        return targetRepository.findByAssignedDistributionSetId(pageReq, distributionSetID);
+        return targetRepository.findByAssignedDistributionSetId(pageReq, distributionSetId);
     }
 
     @Override
-    public Page<Target> findByAssignedDistributionSetAndRsql(final Pageable pageReq, final long distributionSetID,
-            final String rsqlParam) {
-        throwEntityNotFoundIfDsDoesNotExist(distributionSetID);
+    public Page<Target> findByAssignedDistributionSetAndQuery(final Pageable pageable, final long distributionSetId,
+            final String query) {
+        throwEntityNotFoundIfDsDoesNotExist(distributionSetId);
+        // Maybe optimize in order to fetch only the IDs from the database:
+        List<String> controllerIds = targetRepository
+                .findByAssignedDistributionSetId(Pageable.unpaged(), distributionSetId).stream()
+                .map(t -> t.getControllerId()).collect(Collectors.toList());
+        return convertPage(targetQueryExecutionManagement.findByQuery(pageable, query, controllerIds), pageable);
+    }
+
+    // Used by Auto-Assign checker: Identify new targets which have not an
+    // specific
+    // DS assigned...
+    @Override
+    public Page<Target> findByNoActionWithDistributionSetExistsAndQuery(final Pageable pageable,
+            final long distributionSetId, final String query) {
+        throwEntityNotFoundIfDsDoesNotExist(distributionSetId);
+        Set<String> controllerIds = targetRepository.findByNoActionWithDistributionSetExisits(distributionSetId);
+        return convertPage(targetQueryExecutionManagement.findByQuery(pageable, query, controllerIds), pageable);
+    }
 
-        final Specification<JpaTarget> spec = RSQLUtility.parse(rsqlParam, TargetFields.class, virtualPropertyReplacer,
-                database);
+    @Override
+    public long countByNotAssignedDistributionSetAndQuery(final long distributionSetId, final String query) {
+        throwEntityNotFoundIfDsDoesNotExist(distributionSetId);
+        Set<String> targetIds = targetRepository.findByNoActionWithDistributionSetExisits(distributionSetId);
+        return targetQueryExecutionManagement.countByQuery(query, targetIds);
+    }
 
-        return convertPage(
-                targetRepository
-                        .findAll((Specification<JpaTarget>) (root, query,
-                                cb) -> cb.and(TargetSpecifications.hasAssignedDistributionSet(distributionSetID)
-                                        .toPredicate(root, query, cb), spec.toPredicate(root, query, cb)),
-                                pageReq),
-                pageReq);
+    @Override
+    public Page<Target> findByInstalledDistributionSetAndQuery(final Pageable pageable, final long distributionSetId,
+            final String query) {
+        throwEntityNotFoundIfDsDoesNotExist(distributionSetId);
+        // Maybe optimize in order to fetch only the IDs from the database:
```

Comment:
remove comment and apply optimization, for example, use native query in repository to select only ids

Reply:
done